File: Lecture-8-Build-Netflix-Recommendation-System/code/pythonDemo/src/pythondemo/movie_service.py
import json
import math
from pathlib import Path

from .embedding import EmbeddingClient
from .model.movie import Movie
from .model.movie_data import MovieData
from .model.movie_match import MovieMatch

RESOURCES_DIR = Path(__file__).resolve().parent / "resources"
TOP_K_MATCHES = 3


class MovieService:
    def __init__(self, embedding_client: EmbeddingClient):
        self.embedding_client = embedding_client
        self.movies_embedding: list[Movie] = []

    def initialize_movies(self) -> None:
        resource = RESOURCES_DIR / "movies.json"

        with resource.open("r", encoding="utf-8") as input_stream:
            movie_data_list = [
                MovieData(**movie_data) for movie_data in json.load(input_stream)
            ]

        for movie_data in movie_data_list:
            embedding = self.embedding_client.embed(movie_data.description)

            self.movies_embedding.append(
                Movie(
                    title=movie_data.title,
                    description=movie_data.description,
                    embedding=embedding,
                )
            )

        print(f"{len(self.movies_embedding)} movies loaded with embeddings.")

    def search(self, query: str) -> list[MovieMatch]:
        user_query_embedding = self.embedding_client.embed(query)

        matches = [
            MovieMatch(
                title=movie.title,
                description=movie.description,
                match=self._cosine_similarity(user_query_embedding, movie.embedding),
            )
            for movie in self.movies_embedding
        ]

        return self._top_k_matches(self._sort_by_similarity(matches))

    def similar_movies(self, title: str) -> list[MovieMatch]:
        selected_movie = self._find_movie(title)

        matches = [
            MovieMatch(
                title=movie.title,
                description=movie.description,
                match=self._cosine_similarity(
                    selected_movie.embedding, movie.embedding
                ),
            )
            for movie in self.movies_embedding
            if movie.title.lower() != title.lower()
        ]

        return self._top_k_matches(self._sort_by_similarity(matches))

    def _find_movie(self, title: str) -> Movie:
        for movie in self.movies_embedding:
            if movie.title.lower() == title.lower():
                return movie

        raise ValueError(f"Movie not found: {title}")

    @staticmethod
    def _cosine_similarity(a: list[float], b: list[float]) -> float:
        dot_product = sum(x * y for x, y in zip(a, b))
        norm_a = math.sqrt(sum(x * x for x in a))
        norm_b = math.sqrt(sum(y * y for y in b))

        if norm_a == 0 or norm_b == 0:
            return 0.0

        return dot_product / (norm_a * norm_b)

    @staticmethod
    def _sort_by_similarity(matches: list[MovieMatch]) -> list[MovieMatch]:
        return sorted(matches, key=lambda match: match.match, reverse=True)

    @staticmethod
    def _top_k_matches(matches: list[MovieMatch]) -> list[MovieMatch]:
        return matches[:TOP_K_MATCHES]

File: Lecture-8-Build-Netflix-Recommendation-System/code/goDemo/internal/service/movie_service.go
package service

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"os"
	"sort"
	"strings"

	"goDemo/internal/embedding"
	"goDemo/internal/model"
)

const topKMatches = 3

// ErrMovieNotFound is returned by SimilarMovies when the title does not exist in the catalog.
var ErrMovieNotFound = errors.New("movie not found")

type MovieService struct {
	embeddingClient embedding.Client
	moviesEmbedding []model.Movie
}

func NewMovieService(embeddingClient embedding.Client) *MovieService {
	return &MovieService{embeddingClient: embeddingClient}
}

// InitializeMovies loads the catalog from resourcePath and precomputes an embedding for each movie.
func (s *MovieService) InitializeMovies(ctx context.Context, resourcePath string) error {
	data, err := os.ReadFile(resourcePath)
	if err != nil {
		return fmt.Errorf("reading movies resource: %w", err)
	}

	var movieDataList []model.MovieData
	if err := json.Unmarshal(data, &movieDataList); err != nil {
		return fmt.Errorf("parsing movies resource: %w", err)
	}

	for _, movieData := range movieDataList {
		vector, err := s.embeddingClient.Embed(ctx, movieData.Description)
		if err != nil {
			return fmt.Errorf("embedding movie %q: %w", movieData.Title, err)
		}

		s.moviesEmbedding = append(s.moviesEmbedding, model.Movie{
			Title:       movieData.Title,
			Description: movieData.Description,
			Embedding:   vector,
		})
	}

	fmt.Printf("%d movies loaded with embeddings.\n", len(s.moviesEmbedding))
	return nil
}

// Search ranks the catalog against a free-text query embedding.
func (s *MovieService) Search(ctx context.Context, query string) ([]model.MovieMatch, error) {
	userQueryEmbedding, err := s.embeddingClient.Embed(ctx, query)
	if err != nil {
		return nil, err
	}

	matches := make([]model.MovieMatch, 0, len(s.moviesEmbedding))
	for _, movie := range s.moviesEmbedding {
		matches = append(matches, model.MovieMatch{
			Title:       movie.Title,
			Description: movie.Description,
			Match:       cosineSimilarity(userQueryEmbedding, movie.Embedding),
		})
	}

	return topMatches(sortBySimilarity(matches)), nil
}

// SimilarMovies ranks the catalog against the embedding of an existing movie.
func (s *MovieService) SimilarMovies(title string) ([]model.MovieMatch, error) {
	selectedMovie, err := s.findMovie(title)
	if err != nil {
		return nil, err
	}

	matches := make([]model.MovieMatch, 0, len(s.moviesEmbedding))
	for _, movie := range s.moviesEmbedding {
		if strings.EqualFold(movie.Title, title) {
			continue
		}

		matches = append(matches, model.MovieMatch{
			Title:       movie.Title,
			Description: movie.Description,
			Match:       cosineSimilarity(selectedMovie.Embedding, movie.Embedding),
		})
	}

	return topMatches(sortBySimilarity(matches)), nil
}

func (s *MovieService) findMovie(title string) (model.Movie, error) {
	for _, movie := range s.moviesEmbedding {
		if strings.EqualFold(movie.Title, title) {
			return movie, nil
		}
	}

	return model.Movie{}, fmt.Errorf("%w: %s", ErrMovieNotFound, title)
}

func cosineSimilarity(a, b []float64) float64 {
	var dotProduct, normA, normB float64

	for i := range a {
		dotProduct += a[i] * b[i]
		normA += a[i] * a[i]
		normB += b[i] * b[i]
	}

	if normA == 0 || normB == 0 {
		return 0
	}

	return dotProduct / (math.Sqrt(normA) * math.Sqrt(normB))
}

func sortBySimilarity(matches []model.MovieMatch) []model.MovieMatch {
	sort.Slice(matches, func(i, j int) bool {
		return matches[i].Match > matches[j].Match
	})
	return matches
}

func topMatches(matches []model.MovieMatch) []model.MovieMatch {
	if len(matches) > topKMatches {
		return matches[:topKMatches]
	}
	return matches
}
